src/machine.rs

```diff
@@ -1094,6 +1094,44 @@ impl OlmMachine {
         });
     }
 
+    /// Register a callback which will be called whenever a secret
+    /// (`m.secret.send`) is received.
+    ///
+    /// To request a secret from other devices, a client sends an
+    /// `m.secret.request` device event with action set to request and name set
+    /// to the identifier of the secret. A device that wishes to share the
+    /// secret will reply with an m.secret.send event, encrypted using olm.
+    ///
+    /// The secrets are guaranteed to have been received over a 1-to-1 encrypted
+    /// to_device message from a verified own device.
+    ///
+    /// See https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk_crypto/store/struct.Store.html#method.secrets_stream for more information.
+    ///
+    /// `callback` should be a function that takes 2 arguments the secret name
+    /// and value and returns a Promise.
+    #[wasm_bindgen(js_name = "registerReceiveSecretCallback")]
+    pub async fn register_receive_secret_callback(&self, callback: Function) {
+        let stream = self.inner.store().secrets_stream();
+        let store = Arc::new(self.inner.store().clone());
+        let store_weak = Arc::downgrade(&store);
+        // fire up a promise chain which will call `cb` on each result from the stream
+        spawn_local(async move {
+            // Pin the stream to ensure it can be safely moved across threads
+            pin_mut!(stream);
+            while let Some(secret) = stream.next().await {
+                send_secret_gossip_to_callback(&callback, &secret).await;
+                if let Some(store) = store_weak.upgrade() {
+                    match store.delete_secrets_from_inbox(&secret.secret_name).await {
```

Comment:
if the JS callback fails, we will still delete the secrets from the inbox. So we will lose the secret.

It is generally odd to me that registering a callback is enough that the secrets are automatically deleted.

Reply:
Changed, the js layer should now call delete secret.